miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/component/AsyncOperationManager.java

```diff
@@ -0,0 +1,148 @@
+      } else if (service.get(item.getId()) == null) {
+        throw new RestException(String.format("No %s found with ID: %d", type, item.getId()), Status.BAD_REQUEST);
+      }
+      items.add(item);
+    }
+    BulkSaveOperation<R> operation = service.startBulkUpdate(items);
+    String uuid = addAsyncOperation(operation);
+    return makeRunningProgress(uuid, operation);
+  }
+
+  public <T, R extends Identifiable> ObjectNode getAsyncProgress(String uuid, Class<R> itemClass, BulkSaveService<R> service,
+      AuthorizationManager authorizationManager, Function<R, T> toDto) throws Exception {
+    BulkSaveOperation<?> operation = asyncOperations.get(uuid);
+    if (operation == null) {
+      throw new RestException("No operation found with ID: " + uuid, Status.NOT_FOUND);
+    }
+    authorizationManager.throwIfNotOwner(operation.getOwner());
+    if (!itemClass.isAssignableFrom(operation.getItemClass())) {
+      throw new RestException("Invalid operation type", Status.BAD_REQUEST);
+    }
+    if (!operation.isComplete()) {
+      return makeRunningProgress(uuid, operation);
+    } else {
+      if (operation.isSuccess()) {
+        List<T> dtos = service.listByIdList(operation.getSavedIds()).stream()
+            .map(toDto)
+            .collect(Collectors.toList());
+        return makeCompletedProgress(uuid, operation, dtos);
+      } else {
+        return makeFailedProgress(uuid, operation);
+      }
+    }
+  }
+
+  private static ObjectNode makeRunningProgress(String uuid, BulkSaveOperation<?> operation) {
+    return makeBaseProgress(uuid, operation, "running");
+  }
+
+  private static <T> ObjectNode makeCompletedProgress(String uuid, BulkSaveOperation<?> operation, List<T> dtos) {
+    ObjectNode json = makeBaseProgress(uuid, operation, "completed");
+    json.putPOJO("data", dtos);
+    return json;
+  }
+
+  private static ObjectNode makeFailedProgress(String uuid, BulkSaveOperation<?> operation) {
+    ObjectNode json = makeBaseProgress(uuid, operation, "failed");
+    if (operation.getException() instanceof BulkValidationException) {
+      BulkValidationException bulkValidation = (BulkValidationException) operation.getException();
+      json.put("detail", bulkValidation.getLocalizedMessage());
+      RestUtils.addBulkValidationData(json, bulkValidation);
+    } else {
+      json.put("detail", "An unexpected error has occurred");
+    }
+    return json;
+  }
+
+  private static ObjectNode makeBaseProgress(String uuid, BulkSaveOperation<?> operation, String status) {
+    ObjectMapper mapper = new ObjectMapper();
+    ObjectNode json = mapper.createObjectNode();
+    json.put("operationId", uuid);
+    json.put("status", status);
+    json.put("completedUnits", operation.getProgress());
+    json.put("totalUnits", operation.getTotalCount());
+    return json;
+  }
+
+  private <R extends Identifiable> String addAsyncOperation(BulkSaveOperation<R> operation) {
+    String uuid = null;
```

Comment:
You could also allocate a static `AtomicLong` and use `getAndIncrement()`. Not really necessary, but probably faster than even generating a single UUID.